File: code/java/ch14/practice/Challenge03SubarrayDivisibleK.java
package ch14.practice;

import java.util.*;

/**
 * Challenge 3: Subarray Sum Divisible by K
 * Chapter 14: Prefix Sums — The Running Total Trick
 *
 * PROBLEM: Count subarrays with sum divisible by k.
 *
 * EXAMPLES:
 *   solve([4,5,0,-2,-3,1], 5) -> 7
 *   solve([5], 9)             -> 0
 *
 * INSTRUCTIONS: Replace the body of solve() with your solution.
 */
public class Challenge03SubarrayDivisibleK {
    public static int solve(int[] arr, int k) {
        // TODO: Replace this with your solution
        return 0;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int[] arr = Arrays.stream(sc.nextLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        int k = Integer.parseInt(sc.nextLine().trim());
        System.out.println(solve(arr, k));
        sc.close();
    }
}

File: code/python/ch14/practice/challenge_03_subarray_divisible_k.py
"""
Challenge 3: Subarray Sum Divisible by K
==========================================
Chapter 14: Prefix Sums — The Power of Pre-computation

PROBLEM
-------
Given an integer array and an integer k, return the number of contiguous
subarrays whose sum is divisible by k.

INPUT FORMAT
------------
First line: space-separated integers (the array).
Second line: a single integer k.

OUTPUT FORMAT
-------------
A single integer — the count of subarrays with sum divisible by k.

CONSTRAINTS
-----------
- 1 <= len(arr) <= 10^5
- -10^6 <= arr[i] <= 10^6
- 2 <= k <= 10^4

EXAMPLES
--------
Input:
  4 5 0 -2 -3 1
  5
Output: 7

Input:
  5 10 15
  5
Output: 6

INSTRUCTIONS
------------
Replace the `pass` in the solve() function with your solution.
The main block at the bottom handles input/output — don't change it.
"""


def solve(arr: list[int], k: int) -> int:
    """Return count of subarrays with sum divisible by k."""
    pass  # TODO: Replace this with your solution



# ── Do not change anything below this line ──────────────────────────
if __name__ == "__main__":
    arr = list(map(int, input().split()))
    k = int(input())
    print(solve(arr, k))

